File: concepts/05-trees/python/binary_tree.py
"""
Binary Tree Implementation in Python
"""

class TreeNode:
    """Node class for binary tree"""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BinaryTree:
    """Binary Tree Implementation"""
    
    def __init__(self, root=None):
        self.root = root
    
    def inorder_traversal(self, root=None):
        """Inorder: Left, Root, Right - O(n)"""
        if root is None:
            root = self.root
        
        result = []
        
        def inorder(node):
            if node:
                inorder(node.left)
                result.append(node.val)
                inorder(node.right)
        
        inorder(root)
        return result
    
    def preorder_traversal(self, root=None):
        """Preorder: Root, Left, Right - O(n)"""
        if root is None:
            root = self.root
        
        result = []
        
        def preorder(node):
            if node:
                result.append(node.val)
                preorder(node.left)
                preorder(node.right)
        
        preorder(root)
        return result
    
    def postorder_traversal(self, root=None):
        """Postorder: Left, Right, Root - O(n)"""
        if root is None:
            root = self.root
        
        result = []
        
        def postorder(node):
            if node:
                postorder(node.left)
                postorder(node.right)
                result.append(node.val)
        
        postorder(root)
        return result
    
    def level_order_traversal(self, root=None):
        """Level-order (BFS) - O(n)"""
        if root is None:
            root = self.root
        
        if not root:
            return []
        
        from collections import deque
        result = []
        queue = deque([root])
        
        while queue:
            level_size = len(queue)
            level = []
            
            for _ in range(level_size):
                node = queue.popleft()
                level.append(node.val)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            result.append(level)
        
        return result
    
    def max_depth(self, root=None):
        """Maximum depth/height - O(n)"""
        if root is None:
            root = self.root
        
        if not root:
            return 0
        
        return 1 + max(self.max_depth(root.left), self.max_depth(root.right))
    
    def is_balanced(self, root=None):
        """Check if tree is balanced - O(n)"""
        if root is None:
            root = self.root
        
        def check_balance(node):
            if not node:
                return 0, True
            
            left_height, left_balanced = check_balance(node.left)
            right_height, right_balanced = check_balance(node.right)
            
            balanced = (left_balanced and right_balanced and 
                       abs(left_height - right_height) <= 1)
            
            return max(left_height, right_height) + 1, balanced
        
        _, balanced = check_balance(root)
        return balanced


class BinarySearchTree:
    """Binary Search Tree Implementation"""
    
    def __init__(self):
        self.root = None
    
    def insert(self, val):
        """Insert value - O(log n) average, O(n) worst"""
        self.root = self._insert(self.root, val)
    
    def _insert(self, node, val):
        if not node:
            return TreeNode(val)
        
        if val < node.val:
            node.left = self._insert(node.left, val)
        elif val > node.val:
            node.right = self._insert(node.right, val)
        
        return node
    
    def search(self, val):
        """Search for value - O(log n) average, O(n) worst"""
        return self._search(self.root, val)
    
    def _search(self, node, val):
        if not node or node.val == val:
            return node
        
        if val < node.val:
            return self._search(node.left, val)
        return self._search(node.right, val)
    
    def delete(self, val):
        """Delete value - O(log n) average, O(n) worst"""
        self.root = self._delete(self.root, val)
    
    def _delete(self, node, val):
        if not node:
            return node
        
        if val < node.val:
            node.left = self._delete(node.left, val)
        elif val > node.val:
            node.right = self._delete(node.right, val)
        else:
            if not node.left:
                return node.right
            elif not node.right:
                return node.left
            
            # Node with two children
            min_node = self._min_value_node(node.right)
            node.val = min_node.val
            node.right = self._delete(node.right, min_node.val)
        
        return node
    
    def _min_value_node(self, node):
        current = node
        while current.left:
            current = current.left
        return current
    
    def is_valid(self):
        """Validate BST - O(n)"""
        def validate(node, min_val=float('-inf'), max_val=float('inf')):
            if not node:
                return True
            
            if node.val <= min_val or node.val >= max_val:
                return False
            
            return (validate(node.left, min_val, node.val) and
                   validate(node.right, node.val, max_val))
        
        return validate(self.root)
    
    def lowest_common_ancestor(self, p, q):
        """Find LCA of two nodes - O(log n)"""
        node = self.root
        
        while node:
            if p < node.val and q < node.val:
                node = node.left
            elif p > node.val and q > node.val:
                node = node.right
            else:
                return node.val
        
        return None


# Example usage
if __name__ == "__main__":
    # Binary Tree
    print("--- Binary Tree ---")
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    
    tree = BinaryTree(root)
    print("Inorder:", tree.inorder_traversal())
    print("Preorder:", tree.preorder_traversal())
    print("Postorder:", tree.postorder_traversal())
    print("Level-order:", tree.level_order_traversal())
    print("Max depth:", tree.max_depth())
    
    # BST
    print("\n--- Binary Search Tree ---")
    bst = BinarySearchTree()
    for val in [5, 3, 7, 2, 4, 6, 8]:
        bst.insert(val)
    
    print("Inorder:", BinaryTree(bst.root).inorder_traversal())
    print("Search 4:", bst.search(4) is not None)
    print("Is valid BST:", bst.is_valid())
    print("LCA of 2 and 4:", bst.lowest_common_ancestor(2, 4))

File: concepts/05-trees/node/binaryTree.js
/**
 * Binary Tree Implementation in Node.js
 */

class TreeNode {
    /**
     * Node class for binary tree
     */
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class BinaryTree {
    /**
     * Binary Tree Implementation
     */
    constructor(root = null) {
        this.root = root;
    }

    /**
     * Inorder: Left, Root, Right - O(n)
     */
    inorderTraversal(root = null) {
        if (root === null) {
            root = this.root;
        }

        const result = [];

        function inorder(node) {
            if (node) {
                inorder(node.left);
                result.push(node.val);
                inorder(node.right);
            }
        }

        inorder(root);
        return result;
    }

    /**
     * Preorder: Root, Left, Right - O(n)
     */
    preorderTraversal(root = null) {
        if (root === null) {
            root = this.root;
        }

        const result = [];

        function preorder(node) {
            if (node) {
                result.push(node.val);
                preorder(node.left);
                preorder(node.right);
            }
        }

        preorder(root);
        return result;
    }

    /**
     * Postorder: Left, Right, Root - O(n)
     */
    postorderTraversal(root = null) {
        if (root === null) {
            root = this.root;
        }

        const result = [];

        function postorder(node) {
            if (node) {
                postorder(node.left);
                postorder(node.right);
                result.push(node.val);
            }
        }

        postorder(root);
        return result;
    }

    /**
     * Level-order (BFS) - O(n)
     */
    levelOrderTraversal(root = null) {
        if (root === null) {
            root = this.root;
        }

        if (!root) {
            return [];
        }

        const result = [];
        const queue = [root];

        while (queue.length > 0) {
            const levelSize = queue.length;
            const level = [];

            for (let i = 0; i < levelSize; i++) {
                const node = queue.shift();
                level.push(node.val);

                if (node.left) {
                    queue.push(node.left);
                }
                if (node.right) {
                    queue.push(node.right);
                }
            }

            result.push(level);
        }

        return result;
    }

    /**
     * Maximum depth/height - O(n)
     */
    maxDepth(root = null) {
        if (root === null) {
            root = this.root;
        }

        if (!root) {
            return 0;
        }

        return 1 + Math.max(this.maxDepth(root.left), this.maxDepth(root.right));
    }

    /**
     * Check if tree is balanced - O(n)
     */
    isBalanced(root = null) {
        if (root === null) {
            root = this.root;
        }

        function checkBalance(node) {
            if (!node) {
                return [0, true];
            }

            const [leftHeight, leftBalanced] = checkBalance(node.left);
            const [rightHeight, rightBalanced] = checkBalance(node.right);

            const balanced = (leftBalanced && rightBalanced &&
                Math.abs(leftHeight - rightHeight) <= 1);

            return [Math.max(leftHeight, rightHeight) + 1, balanced];
        }

        const [, balanced] = checkBalance(root);
        return balanced;
    }
}

class BinarySearchTree {
    /**
     * Binary Search Tree Implementation
     */
    constructor() {
        this.root = null;
    }

    /**
     * Insert value - O(log n) average, O(n) worst
     */
    insert(val) {
        this.root = this._insert(this.root, val);
    }

    _insert(node, val) {
        if (!node) {
            return new TreeNode(val);
        }

        if (val < node.val) {
            node.left = this._insert(node.left, val);
        } else if (val > node.val) {
            node.right = this._insert(node.right, val);
        }

        return node;
    }

    /**
     * Search for value - O(log n) average, O(n) worst
     */
    search(val) {
        return this._search(this.root, val);
    }

    _search(node, val) {
        if (!node || node.val === val) {
            return node;
        }

        if (val < node.val) {
            return this._search(node.left, val);
        }
        return this._search(node.right, val);
    }

    /**
     * Delete value - O(log n) average, O(n) worst
     */
    delete(val) {
        this.root = this._delete(this.root, val);
    }

    _delete(node, val) {
        if (!node) {
            return node;
        }

        if (val < node.val) {
            node.left = this._delete(node.left, val);
        } else if (val > node.val) {
            node.right = this._delete(node.right, val);
        } else {
            if (!node.left) {
                return node.right;
            } else if (!node.right) {
                return node.left;
            }

            // Node with two children
            const minNode = this._minValueNode(node.right);
            node.val = minNode.val;
            node.right = this._delete(node.right, minNode.val);
        }

        return node;
    }

    _minValueNode(node) {
        let current = node;
        while (current.left) {
            current = current.left;
        }
        return current;
    }

    /**
     * Validate BST - O(n)
     */
    isValid() {
        function validate(node, minVal = -Infinity, maxVal = Infinity) {
            if (!node) {
                return true;
            }

            if (node.val <= minVal || node.val >= maxVal) {
                return false;
            }

            return (validate(node.left, minVal, node.val) &&
                validate(node.right, node.val, maxVal));
        }

        return validate(this.root);
    }

    /**
     * Find LCA of two nodes - O(log n)
     */
    lowestCommonAncestor(p, q) {
        let node = this.root;

        while (node) {
            if (p < node.val && q < node.val) {
                node = node.left;
            } else if (p > node.val && q > node.val) {
                node = node.right;
            } else {
                return node.val;
            }
        }

        return null;
    }
}

// Example usage
if (require.main === module) {
    // Binary Tree
    console.log("--- Binary Tree ---");
    const root = new TreeNode(1);
    root.left = new TreeNode(2);
    root.right = new TreeNode(3);
    root.left.left = new TreeNode(4);
    root.left.right = new TreeNode(5);

    const tree = new BinaryTree(root);
    console.log("Inorder:", tree.inorderTraversal());
    console.log("Preorder:", tree.preorderTraversal());
    console.log("Postorder:", tree.postorderTraversal());
    console.log("Level-order:", tree.levelOrderTraversal());
    console.log("Max depth:", tree.maxDepth());

    // BST
    console.log("\n--- Binary Search Tree ---");
    const bst = new BinarySearchTree();
    [5, 3, 7, 2, 4, 6, 8].forEach(val => bst.insert(val));

    console.log("Inorder:", new BinaryTree(bst.root).inorderTraversal());
    console.log("Search 4:", bst.search(4) !== null);
    console.log("Is valid BST:", bst.isValid());
    console.log("LCA of 2 and 4:", bst.lowestCommonAncestor(2, 4));
}

module.exports = { TreeNode, BinaryTree, BinarySearchTree };

